base/src/com/google/idea/blaze/base/sync/workspace/WorkspaceHelper.java

```diff
@@ -155,31 +183,38 @@ private static WorkspacePath getPackagePath(
     return null;
   }
 
+  @VisibleForTesting
+  public static File getExternalSourceRoot(BlazeProjectData projectData) {
+    return new File(projectData.getBlazeInfo().getOutputBase(), "external");
+  }
+
   @Nullable
-  private static synchronized Map<String, WorkspaceRoot> getExternalWorkspaceRoots(
+  private static synchronized WorkspaceRoot getExternalWorkspaceRootsFile(String workspaceName,
       Project project) {
     if (Blaze.getBuildSystemName(project) == BuildSystemName.Blaze) {
-      return ImmutableMap.of();
+      return null;
     }
-    return SyncCache.getInstance(project)
-        .get(WorkspaceHelper.class, WorkspaceHelper::enumerateExternalWorkspaces);
-  }
-
-  @SuppressWarnings("unused")
-  private static Map<String, WorkspaceRoot> enumerateExternalWorkspaces(
-      Project project, BlazeProjectData blazeProjectData) {
-    FileOperationProvider provider = FileOperationProvider.getInstance();
-    File[] children = provider.listFiles(getExternalSourceRoot(blazeProjectData));
-    if (children == null) {
-      return ImmutableMap.of();
+    logger.debug("getExternalWorkspaceRootsFile for " + workspaceName);
+    File externalBase = SyncCache.getInstance(project)
+        .get(workspaceName, (Project theProject, BlazeProjectData projectData) -> {
```

Comment:
I used the `workspaceName ` as key to store the `workspaceRoot` path. It might be a bit overkill to use it. I also think I could just keep a synchronized map instance in this class. But since the workspaceRoot map before was saved to the SyncCache so I went for it.

Reply:
yeah, but there was a whole map in a single entry for a key named `WorkspaceHelper.class` while after your change there are N entries there.

Reply:
Yeah, the prev implementation cache the entire map with the WorkspaceHelper class name. The map is huge with tens of thousands of entries, most of them is not useful. Because when you use the rule_jvm_external pinned version, all the external dependencies got to download directly into the `external` dir. The map will cache all the external dependencies dir. 

With the name, it will only cache the workspaceName that you installed with the `rules_jvm_external`, with its path. for example `android_mvn`, `test_mvn`...

Reply:
Please check this out 
![image](https://github.com/bazelbuild/intellij/assets/1066003/ee5881c2-77a4-4114-92ec-946a2a01dd49)

Previously, SyncCache had a static number of entries, typically one per purpose, with hardcoded keys in the plugin's codebase. With your change, the cache becomes dynamic, and the number of entries varies.

The risk is that external workspace names are flat-structured. If another service follows your pattern, conflicts may arise. Other services might also need to store external workspace-related data and should reserve their own keys.

cc @ujohnny

Reply:
Ok. I create a map and cache it.

Reply:
I changed the impl to use a local map to cache the workspace root dirs. Using the `SyncCache` does not make sense if we want to cache an map and consistently modifying the entries of the map. This should also fixed the issue you saw below when the init `bazelProjectData` is null? The workspace root will only be cached if the value if the dir existed, means the `bazelProjectData` is not null. Once the project is done initialization if you do the query again the value should returned as expected. @tpasternak